File: app/src/main/java/com/example/nts_pim/utilities/bluetooth_helper/ntsPimPacket.java
package com.example.nts_pim.utilities.bluetooth_helper;

import android.util.Log;

import org.json.JSONObject;

import javax.annotation.Signed;

import kotlin.Pair;
import kotlin.reflect.jvm.internal.impl.util.Checks;


// Right now, all of the commands are 2 characters long.  If this changes, change this constant.
public class ntsPimPacket
{
    // Right now, all of the commands are 2 characters long.  If this changes, change this constant.
    private static final int CMD_BYTE_LENGTH = 2;

    /**
     * Indicates the type of information that should be found in this packet. The strings that represent the command
     * should all have the same number of characters.
     */
    public enum Command
    {
        /**
         * Command to return after receiving a valid packet.
         */
        ACK("OK"),
        /**
         * Command to return if a packet wasn't valid.
         */
        NACK("NO"),
        /**
         * Command containing data for the current status of the Driver App (such as vehicle status, trip info, and meter
         * status).
         */
        MDT_STATUS("MS"),
        /**
         * Command to start payment process on PIM. Should contain all the data needed to start payment.
         */
        START_PAYMENT("SP"),
        /**
         * Command Driver App can send if a payment should be canceled (may not end up using this command).
         */
        CANCEL_PAYMENT("CP"),
        /**
         * Command PIM App should send when a payment is complete (after receipt sent). Should contain payment method
         * (cash or card) and rest of payment details for card payment.
         */
        PIM_PAYMENT("PP"),
        /**
         * Command PIM app should send if a card is declined.  Should contain any available decline message.
         */
        PAYMENT_DECLINED("PD"),
        /**
         * Command PIM app should send when its status (which screen is being displayed) changes or after receiving a
         * STATUS_REQ packet.
         */
        PIM_STATUS("PS"),
        /**
         * Command Driver App can send to request the PIM's status.
         */
        STATUS_REQ("SR"),
        /**
         * Command Driver App can send to PIM to print something (may use in the future if we attach a printer to the PIM).
         */
        PRINT_RECEIPT("PR");

        private final String command;

        Command(String cmd)
        {
            command = cmd;
        }
    }

    /**
     * This variable is set if there are more bytes to parse after finding the end of a packet.
     */
    public int LeftoverIndex;

    private enum ReadState
    {
        STX, CMD, LENGTH, DATA, ETX
    }

    private static final int STX = 0x02;
    private static final int ETX = 0x03;

    private ReadState _readState;
    private Command _cmd;
    private String _strCmd;
    private int _bytesRead, _dataLen;
    private byte[] _data;
    private JSONObject _jsonData;

    /**
     * Constructor to use when getting ready to parse data.
     */
    public ntsPimPacket()
    {
        _readState = ReadState.STX;
        _cmd = null;
        _strCmd = "";
        _bytesRead = 0;
        _dataLen = 0;
        _data = null;
        _jsonData = null;
    }

    /**
     * Constructor to use when getting ready to send a packet.
     *
     * @param cmd  Command indicating what sort of data the packet will contain.
     * @param data JSONObject containing packet data or null if no data should be sent.
     */
    public ntsPimPacket(Command cmd, JSONObject data)
    {
        _cmd = cmd;

        if (data == null) {
            _dataLen = 0;
            _data = null;
            _jsonData = null;
        }
        else {
            _jsonData = data;
            String s = _jsonData.toString();
            _data = s.getBytes();
            _dataLen = _data.length;
        }
    }

    /**
     * Checks to see if any of the bytes passed in is the packet start byte.
     *
     * @param data Array of bytes.
     * @return True if start byte was found.
     */
    public static boolean containsPacketStart(byte[] data)
    {
        if (data == null || data.length == 0) return false;

        // Look for STX.
        for (byte b : data) {
            if (b == STX)
                return true;
        }

        return false;
    }

    /**
     * Gets the data from the packet converted to a JSON object.  If null, packet doesn't have any data.
     *
     * @return JSONObject containing packet data or null if there is no data.
     */
    public JSONObject getPacketData()
    {
        return _jsonData;
    }

    /**
     * Call this function after parseData() returns true to check if the data just parsed is valid.
     *
     * @return True if packet contains valid data.
     */
    public boolean isValidPacket()
    {
        // Check if the command is valid.
        if (_strCmd.equals(Command.ACK.command)) {
            _cmd = Command.ACK;
        }
        else if (_strCmd.equals(Command.NACK.command)) {
            _cmd = Command.NACK;
        }
        else if (_strCmd.equals(Command.MDT_STATUS.command)) {
            _cmd = Command.MDT_STATUS;
        }
        else if (_strCmd.equals(Command.START_PAYMENT.command)) {
            _cmd = Command.START_PAYMENT;
        }
        else if (_strCmd.equals(Command.CANCEL_PAYMENT.command)) {
            _cmd = Command.CANCEL_PAYMENT;
        }
        else if (_strCmd.equals(Command.PIM_PAYMENT.command)) {
            _cmd = Command.PIM_PAYMENT;
        }
        else if (_strCmd.equals(Command.PAYMENT_DECLINED.command)) {
            _cmd = Command.PAYMENT_DECLINED;
        }
        else if (_strCmd.equals(Command.PIM_STATUS.command)) {
            _cmd = Command.PIM_STATUS;
        }
        else if (_strCmd.equals(Command.PRINT_RECEIPT.command)) {
            _cmd = Command.PRINT_RECEIPT;
        }
        else if (_strCmd.equals(Command.STATUS_REQ.command)) {
            _cmd = Command.NACK;
        }
        else {
            // Command not recognized, exit.
           // return false;
        }

      if (_data == null) return true;

        try {
            // First, try to convert the data to a string.
            String s = new String(_data);

            // If an error wasn't thrown, convert the string to a JSON object.
            _jsonData = new JSONObject(s);

            // If this point is reached, then data was converted to JSON and going to assume packet is valid.
            return true;
        }
        catch (Exception e) {
            Log.e(e.toString(), "NtsPimPacket.isValidPacket");
        }

        return false;
    }

    /**
     * Parses the bytes passed in. The byte array may only contain a portion of the packet so keep calling parseData()
     * (on the same class instance) until true is returned.  If true is returned, check variable "LeftoverIndex" to see
     * if it is greater than 0 indicating that the data contains part (or all) of another packet and that parsing should
     * start again at this index.
     *
     * @param data Bytes read from socket connection.
     * @return True if the end of the packet was found.
     */
    public boolean parseData(byte[] data)
    {
        byte b;
        char c;

        //We can just use the buffer.

        for (int i = 0, len = data.length; i < len; i++) {
            b = data[i];

            switch (_readState) {
                case STX:
                    if (b == STX) _readState = ReadState.CMD;
                    break;
                case CMD:
                    c = (char) b;
                    _strCmd += c;
                    _bytesRead++;
                    if (_bytesRead == CMD_BYTE_LENGTH) {
                        _readState = ReadState.LENGTH;
                        _bytesRead = 0;
                        // Use data array to parse length.
                        _data = new byte[2];
                    }
                    break;
                case LENGTH:
                    _data[_bytesRead] = b;
                    _bytesRead++;
                    if (_bytesRead == 2) {
                        // Reset bytes read to start keeping track of data bytes.
                        _bytesRead = 0;
                        // Convert length bytes to integer.  The first byte is the most significant byte (MSB) and the second is
                        // the least significant byte (LSB).  Shift the MSB by 8 bits and do a bitwise OR with LSB.  Bytes, when
                        // cast to an integer, can be negative so do a bitwise AND with 0xFF to make them positive.
                        _dataLen = ((_data[0] & 0xFF) << 8) | (_data[1] & 0xFF);
                        if (_dataLen > 0) {
                            _readState = ReadState.DATA;
                            _data = new byte[_dataLen];
                        }
                        else {
                            _readState = ReadState.ETX;
                            _data = null;
                        }
                    }
                    break;
                case DATA:
                    _data[_bytesRead] = b;
                    _bytesRead++;
                    if (_bytesRead == _dataLen) {
                        _readState = ReadState.ETX;
                    }
                    break;
                case ETX:
                    if (b == ETX) {
                        // If there are any more bytes to parse, set leftover index.
                        if (i < len - 1)
                            LeftoverIndex = i + 1;

                        // Done parsing, return true.
                        return true;
                    }
                    break;
            }
        }

        return false;
    }

    public byte[] toBytes()
    {
        byte[] packet, cmd;
        int i;

        // First, figure out how many bytes are in the command.  Even though right now all of the commands are 2 characters
        // long, this may change in the future so going to leave it variable.
        cmd = _cmd.command.getBytes();

        // Create a byte array for the packet to be sent:
        // - 1 byte for STX
        // - 2 bytes for data length
        // - 1 byte for ETX
        // - plus however many data bytes there are.
        packet = new byte[4 + cmd.length + _dataLen];
        packet[0] = STX;

        // Insert command bytes in to packet array.
        System.arraycopy(cmd, 0, packet, 1, cmd.length);
        i = 1 + cmd.length;

        // Split integer up in to two bytes with the most significant byte first (also know as "big endian" order).
        packet[i] = intToByte(_dataLen >> 8);
        packet[i + 1] = intToByte(_dataLen);
        i += 2;

        // Write data bytes. These should already have been set with the constructor.
        if (_dataLen > 0) {
            System.arraycopy(_data, 0, packet, i, _dataLen);
            i += _dataLen;
        }

        packet[i] = ETX;

        return packet;
    }

    /**
     * Converts the integer's lowest order byte in to a signed byte value.
     *
     * @param i Integer to convert to byte.
     * @return Signed byte.
     */
    private byte intToByte(int i)
    {
        int iTemp = i & 0xFF;

        if (iTemp > 127)
            return (byte) (iTemp - 256);
        return (byte) iTemp;
    }
}

File: app/src/main/java/com/example/nts_pim/utilities/bluetooth_helper/NTSPimPacket.kt
package com.example.nts_pim.utilities.bluetooth_helper

import com.example.nts_pim.PimApplication
import com.example.nts_pim.data.repository.TripDetails
import com.example.nts_pim.data.repository.VehicleTripArrayHolder
import com.example.nts_pim.data.repository.model_objects.trip.Trip
import com.example.nts_pim.receivers.BatteryPowerReceiver
import com.example.nts_pim.utilities.enums.LogEnums
import com.example.nts_pim.utilities.logging_service.LoggerHelper
import org.json.JSONObject

// Right now, all of the commands are 2 characters long.  If this changes, change this constant.
/**
* Helper class for sending and receiving packets between Driver app and NTS PIM app. Format:
* <pre>
*   SDDD...E
*   S = start byte (STX := 0x02)
*   D = variable number of data bytes.  This portion should be JSON formatted.
*   E = end byte (ETX := 0x03)
* </pre>
*/

class NTSPimPacket {
    /**
     * Indicates the type of information that should be found in this packet's data (if there is any beyond command).
     */
    enum class Command(val command: String) {
        ACK("ACK"),
        NACK("NACK"),
        MDT_STATUS("MDT_STATUS"),
        PIM_PAYMENT("PIM_PAYMENT"),
        PAYMENT_DECLINED("PAYMENT_DECLINED"),
        PIM_STATUS("PIM_STATUS"),
        STATUS_REQ("STATUS_REQ"),
        GET_UPFRONT_PRICE("GET_UPFRONT_PRICE"),
        UPFRONT_PRICE("UPFRONT_PRICE"),
        UPDATE_TRIP("UPDATE_TRIP")
    }
    private enum class ReadState {STX}
    private var _readState: ReadState? = null
    var command: Command?
        private set
    var packetData: PimDataObj?
        private set
    private var _index = 0
    private var _data: ByteArray? = null
    private var _readBuffer: ByteArray? = null

    constructor() {
        init()
        command = null
        packetData = null
    }
    constructor(
        cmd: Command?,
        data: PimDataObj?
    ) {
        init()
        command = cmd
        packetData = data
    }
    private fun init() {
        _readState = ReadState.STX
        _index = 0
        _data = null
        _readBuffer = null
    }
    interface PimDataObj {
        fun fromJson(obj: JSONObject)
        fun toJson(): JSONObject
    }

    class MdtStatusObj(mId: Int?, mTripNbr: Int?, mMeterState: String?, mPimNoReceipt: String?, mTripId: String?, mTripStatus: String?, mOwedPrice: Double?, mPimPayAmt: Double?) : PimDataObj {
        var driverId: Int? = null
        var tripNbr: Int? = null
        var meterState: String? = null
        var pimNoReceipt: String? = null
        var tripId: String? = null
        var tripStatus: String? = null
        var owedPrice: Double? = null
        var pimPayAmt: Double? = null
        init {
            driverId = mId
            tripNbr = mTripNbr
            meterState = mMeterState
            pimNoReceipt = mPimNoReceipt
            tripId = mTripId
            tripStatus = mTripStatus
            owedPrice = mOwedPrice
            pimPayAmt = mPimPayAmt
        }

        override fun fromJson(obj: JSONObject) {
            tripId = obj.optString(JSON_TRIP_ID)
            pimPayAmt = obj.optDouble(JSON_PIM_PAY_AMT)
            tripNbr = obj.optInt(JSON_TRIP_NBR)
            pimNoReceipt = obj.optString(JSON_PIM_NO_RECEIPT)
            tripStatus = obj.optString(JSON_TRIP_STATUS)
            owedPrice = obj.optDouble(JSON_OWED_PRICE)
            driverId = obj.optInt(JSON_DRIVER_ID)
            meterState = obj.optString(JSON_METER_STATE)
            LoggerHelper.writeToLog("Parsed packet from Json: tripId: $tripId, pimPayAmt: $pimPayAmt,tripNum: $tripNbr, pimNoReceipt: $pimNoReceipt, tripStatus: $tripStatus, owedPrice: $owedPrice, driverId: $driverId, meterState: $meterState", LogEnums.BLUETOOTH.tag)
            if(!tripId.isNullOrBlank() || !tripId.isNullOrEmpty()){
                val thisForTheCurrentTrip = TripDetails.isThisForAnOldTrip(tripId!!)
                if(!thisForTheCurrentTrip){
                    VehicleTripArrayHolder.addTripId(tripId!!, PimApplication.pimContext)
                } else {
                    LoggerHelper.writeToLog("Packet received for $tripId, but was found to be a completed trip. This packet is being ignored", LogEnums.TRIP_STATUS.tag)
                    return
                }
            }
            if(pimPayAmt != null && !pimPayAmt!!.isNaN()){
                VehicleTripArrayHolder.setPimPayment(pimPayAmt!!)
            } else {
                LoggerHelper.writeToLog("pim payamount: $pimPayAmt, it was ignored", LogEnums.BLUETOOTH.tag)
            }

            if(tripNbr != null){
                VehicleTripArrayHolder.addTripNumber(tripNbr!!)
            }

            if(pimNoReceipt != null && pimNoReceipt == "Y"){
                VehicleTripArrayHolder.pimDoesNotNeedToDoReceipt(true)
            }

            if(tripStatus != null){
                VehicleTripArrayHolder.addStatus(tripStatus!!)
            }
            if(owedPrice != null){}
            if(driverId != null){
                VehicleTripArrayHolder.setDriverId(driverId!!)
            }
            if(meterState != null){
                VehicleTripArrayHolder.addMeterState(meterState!!)
            }
        }

        override fun toJson(): JSONObject {
            val obj = JSONObject()
            try {
                if (driverId != null) {obj.put(JSON_DRIVER_ID, driverId!!)}
                if (meterState != null) {obj.put(JSON_METER_STATE, meterState!!)}
                if (pimNoReceipt != null) obj.put(
                    JSON_PIM_NO_RECEIPT,
                    pimNoReceipt!!
                )
                if (tripNbr != null) {obj.put(JSON_TRIP_NBR, tripNbr!!)}
                if (tripId != null) {obj.put(JSON_TRIP_ID, tripId!!)}
                if (tripStatus != null) {obj.put(JSON_TRIP_STATUS, tripStatus!!)}
                if (owedPrice != null) {obj.put(JSON_OWED_PRICE, owedPrice!!)}
                if (pimPayAmt != null) {obj.put(JSON_PIM_PAY_AMT, pimPayAmt!!)}
            } catch (ignore: Exception) {
            }
            return obj
        }

        companion object {
            private const val JSON_DRIVER_ID = "driverId"
            private const val JSON_TRIP_NBR = "tripNbr"
            private const val JSON_METER_STATE = "meterState"
            private const val JSON_PIM_NO_RECEIPT = "pimNoReceipt"
            private const val JSON_TRIP_ID = "tripId"
            private const val JSON_TRIP_STATUS = "tripStatus"
            private const val JSON_OWED_PRICE = "owedPrice"
            private const val JSON_PIM_PAY_AMT = "pimPayAmt"
        }
    }

    /**
     * Class to help send or parse JSON data for PIM Payment packet.  Sent by PIM after a card or cash payment.
     * Trip ID should be included to make sure the payment is recorded for the correct trip (like in the case
     * where a trip is closed during payment process).
     */
    class PimPaymentObj : PimDataObj {
        var pimPaidAmt: Double? = null
        var tipAmt: Double? = null
        var cardInfo: String? = null
       private var payType: String? = null
        var transDate: String? = null
        var transId: String? = null
        var tripId: String? = null
        override fun fromJson(obj: JSONObject) {
            pimPaidAmt = obj.optDouble(JSON_PAID_AMT, 0.0)
            tipAmt = obj.optDouble(JSON_TIP_AMT, 0.0)
            cardInfo = obj.optString(JSON_CARD_INFO)
            payType = obj.optString(JSON_PAY_TYPE)
            transDate = obj.optString(JSON_TRANS_DATE)
            transId = obj.optString(JSON_TRANS_ID)
            tripId = obj.optString(JSON_TRIP_ID)
        }

        override fun toJson(): JSONObject {
            val obj = JSONObject()
            pimPaidAmt = VehicleTripArrayHolder.getPimPaidAmount()
            tipAmt = VehicleTripArrayHolder.getTipAmount()
            cardInfo = VehicleTripArrayHolder.getCardInfo()
            transDate = VehicleTripArrayHolder.getTransDate()
            payType = VehicleTripArrayHolder.paymentTypeSelected
            transId = VehicleTripArrayHolder.getTransactionID()
            tripId = VehicleTripArrayHolder.getTripId()
            try {
                if (pimPaidAmt != null) {obj.put(JSON_PAID_AMT, pimPaidAmt!!)}
                if (tipAmt != null) {obj.put(JSON_TIP_AMT, tipAmt!!)}
                if (cardInfo != null) {obj.put(JSON_CARD_INFO, cardInfo!!)}
                if (payType != null) {obj.put(JSON_PAY_TYPE, payType!!)}
                if (transDate != null) {obj.put(JSON_TRANS_DATE, transDate!!)}
                if (transId != null) {obj.put(JSON_TRANS_ID, transId!!)}
                if (tripId != null) {obj.put(JSON_TRIP_ID, tripId!!)}
            } catch (ignore: Exception) {
            }
            return obj
        }

        companion object {
            private const val JSON_PAID_AMT = "pimPaidAmt"
            private const val JSON_TIP_AMT = "tipAmt"
            private const val JSON_CARD_INFO = "cardInfo"
            private const val JSON_PAY_TYPE = "payType"
            private const val JSON_TRANS_DATE = "transDate"
            private const val JSON_TRANS_ID = "transId"
            private const val JSON_TRIP_ID = "tripId"
        }

    } // end of class PimPaymentObj

    /**
     * Class to help send or parse JSON data for Payment Declined packet.  Sent by PIM so driver can see the same
     * error message as the passenger if a payment is declined.
     */
    class PaymentDeclinedObj() : PimDataObj {
        private var declineMsg: String? = null
        override fun fromJson(obj: JSONObject) {
            declineMsg = obj.optString(JSON_DECLINE_MSG)
        }

        override fun toJson(): JSONObject {
            val obj = JSONObject()
            declineMsg = VehicleTripArrayHolder.getDeclinedCardMessage()
            try {
                if (declineMsg != null) obj.put(
                    JSON_DECLINE_MSG,
                    declineMsg
                )
            } catch (ignore: Exception) {
            }
            return obj
        }

        companion object {
            private const val JSON_DECLINE_MSG = "declineMsg"
        }

    } // end of class PaymentDeclinedObj

    /**
     * Class to help send or parse JSON data for PIM Status packet.  Sent by PIM when its status changes or in
     * response to a Status Request.
     */

    class PimStatusObj : PimDataObj {
        var pimStatus: String? = null
        var tripId: String? = null
        var pimOverHeat: Boolean? = null

        override fun fromJson(obj: JSONObject) {
            tripId = obj.optString(JSON_TRIP_ID)
            pimStatus = obj.optString(JSON_PIM_STATUS)
            pimOverHeat = obj.optBoolean(JSON_OVER_HEAT)
        }

        override fun toJson(): JSONObject {
            val obj = JSONObject()
            pimStatus = VehicleTripArrayHolder.getInternalPIMStatus()
            tripId = VehicleTripArrayHolder.getTripId()
            pimOverHeat = BatteryPowerReceiver.overHeating
            try {
                if (pimStatus != null){
                    obj.put(JSON_PIM_STATUS, pimStatus)
                }
                if(tripId != null && tripId != ""){
                    obj.put(JSON_TRIP_ID, tripId)
                }
                if(pimOverHeat != null){
                    obj.put(JSON_OVER_HEAT, pimOverHeat!!)
                }
            } catch (ignore: Exception) {
            }
            return obj
        }

        companion object {
            private const val JSON_PIM_STATUS = "pimStatus"
            private const val JSON_TRIP_ID = "tripId"
            private const val JSON_OVER_HEAT = "pimOverheat"
        }

    } // end of class PimStatusObj

    companion object {
        private const val JSON_COMMAND = "Command"
        private const val JSON_DATA = "Data"
        private const val STX = 0x02
        private const val ETX = 0x03

        /**
         * Checks to see if any of the bytes passed in is the packet start byte.
         *
         * @param data Array of bytes.
         * @return True if start byte was found.
         */
        fun containsPacketStart(data: ByteArray?): Boolean {
            if (data == null || data.isEmpty()){
                LoggerHelper.writeToLog("Received a packet, but data was null or empty", LogEnums.BLUETOOTH.tag)
                return false
            }
            // Look for STX.
            for (b in data) {
                if (b.toInt() == STX) {
                    return true
                }
            }
            LoggerHelper.writeToLog("Received a packet, but didn't contain start byte", LogEnums.BLUETOOTH.tag)
            return false
        }
    }
}

